### ISSUE TITLE
Don't implement allow-newer/allow-older by modifying PackageDescription

### ISSUE THREAD

**Issue by ezyang** [haskell/cabal#4203]
It seems cleaner to hook it straight into the dep solver, rather than try to ninja it in by package description modification. If you do it by modification, you have to worry about making sure the "correct" package description is used anywhere you want to look at deps, and sometimes you get spurious warnings because it looks like you modify build-depends which affects our source level checks, but actually we didn't.

The lack of cleanliness here is hurting me because I want BuildInfo to reflect the written syntax as closely as possible, but I have to keep `targetBuildDepends` around as a real field if people are editing it.

Previously @kosmikus requested this in #3466.

**Comment by Ericson2314** [comment]
This would be in the same spirit as removing cabal-install's other extra bounds which cause https://github.com/haskell/cabal/issues/4169 right?

**Comment by ezyang** [comment]
Yep, I think so, thanks for posting the reference!

**Comment by 23Skidoo** [comment]
The problem here is that we won't be able to implement `./Setup.hs configure --alow-newer` in this way, and IIRC that interface is used by Stack (it was added per @snoyberg's request).

**Comment by hvr** [comment]
@23Skidoo relevant: https://github.com/haskell/cabal/issues/3581

**Comment by 23Skidoo** [comment]
OK, if Stack doesn't need that interface, then we can remove it.

**Comment by Ericson2314** [comment]
Because of https://github.com/haskell/cabal/pull/4527, this only covers other cases of modifying bounds in situ like #4169, and #4383 if its merged, right?

**Comment by 23Skidoo** [comment]
#4527 only removed the `--allow-{newer,older}` command-line options from setup scripts, the main issue this ticket is about is still there.

**Comment by Ericson2314** [comment]
@23Skidoo Ah the code in Cabal is still there and not dead via something else?

**Comment by 23Skidoo** [comment]
Yes, it's spread across `lib:Cabal` and `cabal-install`. See [`D.C.Dependency.removeBounds`](https://github.com/haskell/cabal/blob/3401b6ba570a0b92979af95400a1749c74faffe1/cabal-install/Distribution/Client/Dependency.hs#L436), which uses [`D.PD.transformAllBuildDepends`](https://github.com/haskell/cabal/blob/66fafcae6f15abc8efe74c191c3c028aa378ebd2/Cabal/Distribution/PackageDescription/Configuration.hs#L616).

**Comment by Ericson2314** [comment]
Thanks!
